### ISSUE TITLE
BUG: scipy.linalg.eigh segfault on MacOS

### ISSUE THREAD

**Issue by sunqm** [scipy/scipy#16151]
### Describe your issue.

On MacOS, scipy.linalg.eigh occasionally crashes with segmentation fault error. The arrays are 96x96 complex128 hermitian arrays.
[hs.npy.tar.gz](https://github.com/scipy/scipy/files/8655144/hs.npy.tar.gz)

The test fails on scipy 1.2.0 - 1.7.3 but works fine if downgrading to scipy 1.1.0. scipy library was installed with pip.

My python version is 3.7.4, numpy version 1.21.2. blas/lapack versions are
```
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/usr/local/lib']
```

### Reproducing Code Example

```python
import numpy as np
import scipy.linalg
h = np.load('h.npy')
s = np.load('s.npy')
scipy.linalg.eigh(h, s)
```

### Error message

```shell
segmentation fault  python hs.py
```

### SciPy/NumPy/Python version information

1.2.0 - 1.7.3

**Comment by ilayn** [comment]
I can't reproduce this on Windows machine with 1.7.3 as it gives me the eigenvalues 

```python
(array([-3.76382203e+04, -3.76382203e+04, -3.75865659e+04, -3.75865659e+04,
        -3.75684694e+04, -3.75684694e+04, -3.75673695e+04, -3.75673695e+04,
        -3.75670502e+04, -3.75670502e+04, -3.75657302e+04, -3.75657302e+04,
        -3.75653710e+04, -3.75653710e+04, -3.75634776e+04, -3.75634776e+04,
        -3.75630807e+04, -3.75630807e+04, -3.75628159e+04, -3.75628159e+04,
        -3.75615751e+04, -3.75615751e+04, -3.75612216e+04, -3.75612216e+04,
        -3.75608895e+04, -3.75608895e+04, -3.75605757e+04, -3.75605757e+04,
        -3.75601823e+04, -3.75601823e+04, -3.75600906e+04, -3.75600906e+04,
        -3.75598683e+04, -3.75598683e+04, -3.75597868e+04, -3.75597868e+04,
        -3.75593463e+04, -3.75593463e+04, -3.75588260e+04, -3.75588260e+04,
        -3.75586004e+04, -3.75586004e+04, -3.75584206e+04, -3.75584206e+04,
        -3.75580322e+04, -3.75580322e+04, -3.75578982e+04, -3.75578982e+04,
        -2.05719271e+01, -2.05719271e+01, -1.33801090e+00, -1.33801090e+00,
        -6.98582055e-01, -6.98582055e-01, -5.66575606e-01, -5.66575606e-01,
        -4.92958379e-01, -4.92958379e-01,  1.85234481e-01,  1.85234481e-01,
         2.56009556e-01,  2.56009556e-01,  7.88939011e-01,  7.88939011e-01,
         8.54600752e-01,  8.54600752e-01,  1.16426231e+00,  1.16426231e+00,
         1.20109416e+00,  1.20109416e+00,  1.25438286e+00,  1.25438286e+00,
         1.44592183e+00,  1.44592183e+00,  1.47899399e+00,  1.47899399e+00,
         1.67696484e+00,  1.67696484e+00,  1.87158201e+00,  1.87158201e+00,
         1.93706006e+00,  1.93706006e+00,  2.45196265e+00,  2.45196265e+00,
         2.49043137e+00,  2.49043137e+00,  3.30125228e+00,  3.30125228e+00,
         3.35106816e+00,  3.35106816e+00,  3.52537859e+00,  3.52537859e+00,
         3.87839957e+00,  3.87839957e+00,  4.15921524e+00,  4.15921524e+00]),
     array[.....]) #  eigvectors array snipped
```
do you mind adding details such as if you are using pip or conda or the result of `scipy.linalg.lapack.ilaver` and other relevant details?

**Comment by matthew-brett** [comment]
I don't get a segfault - with Numpy 1.21.2, and Scipy 1.8.0, installed from Pip - on macOS 11.6.5.

**Comment by sunqm** [comment]
> I can't reproduce this on Windows machine with 1.7.3 as it gives me the eigenvalues
> 
> ```python
> (array([-3.76382203e+04, -3.76382203e+04, -3.75865659e+04, -3.75865659e+04,
>         -3.75684694e+04, -3.75684694e+04, -3.75673695e+04, -3.75673695e+04,
>         -3.75670502e+04, -3.75670502e+04, -3.75657302e+04, -3.75657302e+04,
>         -3.75653710e+04, -3.75653710e+04, -3.75634776e+04, -3.75634776e+04,
>         -3.75630807e+04, -3.75630807e+04, -3.75628159e+04, -3.75628159e+04,
>         -3.75615751e+04, -3.75615751e+04, -3.75612216e+04, -3.75612216e+04,
>         -3.75608895e+04, -3.75608895e+04, -3.75605757e+04, -3.75605757e+04,
>         -3.75601823e+04, -3.75601823e+04, -3.75600906e+04, -3.75600906e+04,
>         -3.75598683e+04, -3.75598683e+04, -3.75597868e+04, -3.75597868e+04,
>         -3.75593463e+04, -3.75593463e+04, -3.75588260e+04, -3.75588260e+04,
>         -3.75586004e+04, -3.75586004e+04, -3.75584206e+04, -3.75584206e+04,
>         -3.75580322e+04, -3.75580322e+04, -3.75578982e+04, -3.75578982e+04,
>         -2.05719271e+01, -2.05719271e+01, -1.33801090e+00, -1.33801090e+00,
>         -6.98582055e-01, -6.98582055e-01, -5.66575606e-01, -5.66575606e-01,
>         -4.92958379e-01, -4.92958379e-01,  1.85234481e-01,  1.85234481e-01,
>          2.56009556e-01,  2.56009556e-01,  7.88939011e-01,  7.88939011e-01,
>          8.54600752e-01,  8.54600752e-01,  1.16426231e+00,  1.16426231e+00,
>          1.20109416e+00,  1.20109416e+00,  1.25438286e+00,  1.25438286e+00,
>          1.44592183e+00,  1.44592183e+00,  1.47899399e+00,  1.47899399e+00,
>          1.67696484e+00,  1.67696484e+00,  1.87158201e+00,  1.87158201e+00,
>          1.93706006e+00,  1.93706006e+00,  2.45196265e+00,  2.45196265e+00,
>          2.49043137e+00,  2.49043137e+00,  3.30125228e+00,  3.30125228e+00,
>          3.35106816e+00,  3.35106816e+00,  3.52537859e+00,  3.52537859e+00,
>          3.87839957e+00,  3.87839957e+00,  4.15921524e+00,  4.15921524e+00]),
>      array[.....]) #  eigvectors array snipped
> ```
> 
> do you mind adding details such as if you are using pip or conda or the result of `scipy.linalg.lapack.ilaver` and other relevant details?
> 
> Also I forgot to check but are you sure `s` is positive definite? Shouldn't lead to a segfault but worth checking

I only observe problems on macos.  I observed the failure on my laptop and the macos virtual machines provided by github actions. My laptop runs python3.7 from miniconda. All other libraries are installed with pip. On github actions, os is set to `macos-latest` with python provided by `actions/checkout@v2`.

s is positive definite.

**Comment by ilayn** [comment]
You can't mix and match conda and pip installs though I am not sure if you are doing that. Also conda uses MKL library which had a LAPACK bug regarding the underlying routines. Other than that, I don't think how to reproduce this as our CI jobs also use MacOS and same GH actions.

**Comment by sunqm** [comment]
MKL and conda are irrelevant here. I've checked the libraries that numpy and scipy are linked to. This error was caught only on scipy 1.2.0 and newer version. It works alright if downgrading to scipy 1.1.0 while keeping everything else the same

I suspect this issue https://github.com/scipy/scipy/issues/15362 is relevant

**Comment by ilayn** [comment]
That one is about eigh(a) and you are using eigh(a, b) which travels a different path. So once again, did you mix and match conda and pip installed libraries? Also did you explicitly select openblas in your miniconda env because the default is MKL?

**Comment by rgommers** [comment]
@sunqm can you please add:
- macOS version
- hardware of your laptop: `x86_64`, `arm64` (and if the latter, are you running `x86_64` binaries?)
- the full output of `conda list` and `conda info`

**Comment by rgommers** [comment]
> You can't mix and match conda and pip installs though I am not sure if you are doing that.

This should actually work fine the way @sunqm describe it - if you _only_ use `conda` to install Python itself, and install everything else from PyPI it should work (still not recommended, but I'm not aware of any immediate problems).

**Comment by sunqm** [comment]
> @sunqm can you please add:
> 
> * macOS version
> * hardware of your laptop: `x86_64`, `arm64` (and if the latter, are you running `x86_64` binaries?)
> * the full output of `conda list` and `conda info`

my laptop:
```
$ sw_vers      
ProductName:    Mac OS X
ProductVersion: 10.15.7
BuildVersion:   19H2
```
```
$ uname -isromv
Darwin 19.6.0 Darwin Kernel Version 19.6.0: Mon Aug 31 22:12:52 PDT 2020; root:xnu-6153.141.2~1/RELEASE_X86_64 x86_64 MacBookPro15,4 Darwin
```
```
$ conda list
awscli                    1.20.6                   pypi_0    pypi
botocore                  1.21.6                   pypi_0    pypi
brotlipy                  0.7.0           py37h9ed2024_1003  
ca-certificates           2021.7.5             hecd8cb5_1  
cached-property           1.5.2                    pypi_0    pypi
certifi                   2021.5.30        py37hecd8cb5_0  
cffi                      1.14.6           py37h2125817_0  
chardet                   4.0.0           py37hecd8cb5_1003  
colorama                  0.4.3                    pypi_0    pypi
conda                     4.10.3           py37hecd8cb5_0  
conda-package-handling    1.7.3            py37h9ed2024_1  
cryptography              3.4.7            py37h2fd3fbb_0  
docutils                  0.15.2                   pypi_0    pypi
h5py                      3.6.0                    pypi_0    pypi
idna                      2.10               pyhd3eb1b0_0  
jmespath                  0.10.0                   pypi_0    pypi
libcxx                    10.0.0                        1  
libffi                    3.3                  hb1e8313_2  
ncurses                   6.2                  h0a44026_1  
numpy                     1.21.6                   pypi_0    pypi
openssl                   1.1.1k               h9ed2024_0  
pandas                    0.25.3                   pypi_0    pypi
pip                       20.0.1                   pypi_0    pypi
pyasn1                    0.4.8                    pypi_0    pypi
pycosat                   0.6.3            py37h9ed2024_0  
pycparser                 2.20                       py_2  
pyopenssl                 20.0.1             pyhd3eb1b0_1  
pysocks                   1.7.1            py37hecd8cb5_0  
python                    3.7.10               h88f2d9e_0  
python-dateutil           2.8.2                    pypi_0    pypi
python.app                3                py37h9ed2024_0  
pyyaml                    5.4.1                    pypi_0    pypi
readline                  8.1                  h9ed2024_0  
requests                  2.25.1             pyhd3eb1b0_0  
rsa                       4.7.2                    pypi_0    pypi
ruamel_yaml               0.15.100         py37h9ed2024_0  
s3transfer                0.5.0                    pypi_0    pypi
scipy                     1.1.0                    pypi_0    pypi
setuptools                52.0.0           py37hecd8cb5_0  
six                       1.16.0             pyhd3eb1b0_0  
sqlite                    3.36.0               hce871da_0  
tk                        8.6.10               hb0a8c7a_0  
tlclient                  0.6.53                   pypi_0    pypi
tqdm                      4.61.2             pyhd3eb1b0_1  
urllib3                   1.26.6             pyhd3eb1b0_1  
wheel                     0.36.2             pyhd3eb1b0_0  
xz                        5.2.5                h1de35cc_0  
yaml                      0.2.5                haf1e3a3_0  
zlib                      1.2.11               h1de35cc_3  
```

```
$ conda info

     active environment : None
       user config file : /Users/qiming.sun/.condarc
 populated config files : 
          conda version : 4.10.3
    conda-build version : not installed
         python version : 3.7.10.final.0
       virtual packages : __osx=10.15.7=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /Users/qiming.sun/opt/miniconda3  (writable)
      conda av data dir : /Users/qiming.sun/opt/miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /Users/qiming.sun/opt/miniconda3/pkgs
                          /Users/qiming.sun/.conda/pkgs
       envs directories : /Users/qiming.sun/opt/miniconda3/envs
                          /Users/qiming.sun/.conda/envs
               platform : osx-64
             user-agent : conda/4.10.3 requests/2.25.1 CPython/3.7.10 Darwin/19.6.0 OSX/10.15.7
                UID:GID : 501:20
             netrc file : None
           offline mode : False
```

**Comment by rgommers** [comment]
Thanks @sunqm, that all looks as it should, so this should be reproducible now at least on a `x86_64` Macbook.

> The test fails on scipy 1.2.0 - 1.7.3

Last question, given that you explicitly exclude 1.8.0 here - did you test with 1.8.0? It ships with a newer version of OpenBLAS IIRC, so perhaps the issue is solved already.

**Comment by sunqm** [comment]
scipy 1.8.0 does not have a wheel for python 3.7. So I installed python 3.9. It does not crash in this test. But I still got segmentation fault for scipy.linalg.eigh in other jobs. I noticed scipy shipped with libopenblas since 1.2.0. Is it possible that the openblas library in scipy has compatibility issue?

**Comment by rgommers** [comment]
> I noticed scipy shipped with libopenblas since 1.2.0. Is it possible that the openblas library in scipy has compatibility issue?

Ah good catch - there are wheels for `1.1.0` but they have no bundled BLAS library at all so there we just expected users to have something installed by the system. So OpenBLAS is crashing here independent of SciPy version. This isn't unusual, we've seen quite a few segfault with it before. 

The next step to make progress here would be to extract a standalone reproducer (i.e., not depending on `.npy` files nor on `scipy` itself, just plain C) and then file an issue on https://github.com/xianyi/OpenBLAS/.

**Comment by steppi** [comment]
@sunqm, since the reproducer above no longer fails for more recent versions of SciPy, can you give another example that still segfaults? After that, I can take care of writing a C only reproducer.

**Comment by steppi** [comment]
Looking into this further; it looks like the original segfault was due to the issue https://github.com/xianyi/OpenBLAS/issues/3309, which was fixed in PR https://github.com/xianyi/OpenBLAS/pull/3311, which was incorporated in OpenBLAS 0.3.17 with SciPy 1.8 updating the shipped OpenBLAS to 0.3.17.

@sunqm, please let me know if you have other examples where `scipy.linalg.eigh(a, b)` segfaults on MacOS.

**Comment by sunqm** [comment]
Updating to the latest scipy solves the problem. Thanks for the fix!
